### ISSUE TITLE
Implement FIFO Support

### ISSUE THREAD

**Issue by tomazfernandes** [awspring/spring-cloud-aws#436]
Implement support for `FIFO` queues.

Should be pretty straightforward overall. The key part should be implementing a `FifoMessageSplitter` with the logic to group messages from the same `message group` and processing each group in parallel while maintaining order within a group.

The `AbstractEndpoint` should be aware that it's a `fifo` endpoint and instantiate the proper `splitter`. Maybe it's worth having a `FifoEndpoint` subclass.

----

The `AsyncMessageSplitter` interface is responsible for splitting the batch of messages acquired from the `AsyncMessagePoller` and sending the messages through a processing pipeline provided by the listener container.

```java
@FunctionalInterface
public interface AsyncMessageSplitter<T> {

	Collection<CompletableFuture<Void>> splitAndProcess(Collection<Message<T>> messages, Function<Message<T>, CompletableFuture<Void>> processingPipeline);
}
```

**Comment by tomazfernandes** [comment]
Just as a note, it seems AWS's [recommended solution](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/FIFO-queues.html) for finding out if a queue is `FIFO` is actually checking the name end for `.fifo`

`Note:
The name of a FIFO queue must end with the .fifo suffix. The suffix counts towards the 80-character queue name quota. To determine whether a queue is FIFO, you can check whether the queue name ends with the suffix.`

**Comment by mgusiew-guide** [comment]
One thing that can be taken under consideration is visibility timeout. In case when one gets 10 messages from same group, they all need to be completed before within the time limit. The idea is to allow to extend the time limit after every message in the group is processed.

The other interesting thing is to decide what to do in case if the message is processed with error. There are few options: 
1) Process next messages in group
2) Retry the error message (possibly with extending timeout on all of messages in the group)
3) Let the messages return to queue (no ack) - the problem here is that until few retries the messages may land up in DLQ even though they were not ever processed (because of poison message in front)

It may happen that some tradeoffs will need to be made and that in order to get expected FIFO semantics, the end user will need to keep batch size set to 1 and perhaps decrease poll timeout.  

IMHO it is bit unfortunate that SQS does not allow to set poll policy to not return multiple messages from same message group as this would allow to parallelize easily (no ordering conflicts in a polled batch)

**Comment by maciejwalkowiak** [comment]
Side note - if possible, we can keep FIFO support for another PR.

**Comment by tomazfernandes** [comment]
Sure @maciejwalkowiak, thanks. Although, considering this is a complex issue, I'm considering picking this up next so we have time to let the solution mature for a while until we release `M2`. But I could pick up the other issues first too. Let me know if you have any further thoughts. Thanks.

**Comment by tomazfernandes** [comment]
I mean, I could open another PR for it anyway, but unless it’s a PR against my branch we’d not be able to see much. Perhaps I didn’t get your point?

**Comment by tomazfernandes** [comment]
I've created a new issue that makes this scope more clear, so I'm closing this one.

Thanks!

Closed in favor of https://github.com/awspring/spring-cloud-aws/issues/444
